### PR TITLE
feat(start): start selenium without making web requests

### PR DESCRIPTION
**Pull request by cnishina**

### CODE REVIEW

**Comment by cnishina** [comment]
Craig please write tests. Also please put the state in `Config` instead of passing it down via `validateCache`. Thanks.

**Comment by cnishina** [comment]
Use `Conig` with `runCommand`. Still needs tests.

**Comment by cnishina** [comment]
Tests added.

File: lib/binaries/config_source.ts [hunk 1]
```diff
@@ -35,24 +35,29 @@ export abstract class XmlConfigSource extends ConfigSource {
     let content = this.readResponse();
     if (content != null) {
       return Promise.resolve(content);
+    } else {
```
Comment (heathkit) [inline lib/binaries/config_source.ts]: I actually slightly prefer the way this used to be, but either is fine.

File: spec/binaries/config_source_spec.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      jsonConfig.testGetJson()
+          .then(json => {
+            expect(json.foo).toEqual('bar');
+          })
+          .catch(err => {
+            fail(err);
+          });
+    });
+
+    it('on udpate: should check the timestamp, invaidate cache, and try to make a web request',
```
Comment (heathkit) [inline spec/binaries/config_source_spec.ts]: Don't you need to be using `(done) =>` in these?

File: lib/binaries/config_source.ts (hunk 1 above)
Comment (cnishina) [inline lib/binaries/config_source.ts]: Agree. Changing to 
```
if(content) {
} else {
}
```

**Comment by attila123** [comment]
Hi @cnishina , @heathkit 
Unfortunately webdriver-manager still cannot work offline.
Jenkins started it as:
```
webdriver-manager start --versions.standalone 3.3.1 --versions.chrome 2.29 --chrome_logs /path/to/chromedriver_7270.log --seleniumPort 7270
```
and now in case of a network failure I got the following error:
```
webdriver-manager: using global installed version 12.0.6
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: getaddrinfo EAI_AGAIN chromedriver.storage.googleapis.com:443
    at Object.exports._errnoException (util.js:1022:11)
    at errnoException (dns.js:33:15)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:76:26)
```
Are you aware of this? Is it supposed to work fully offline with this fix?
(Under Linux (Centos), Node.js version is 6.9.5 if that matters.)

**Comment by anjanapai2508** [comment]
Hi I am trying to run protractor test scripts on an iOS device from my mac and if I do `webdriver-manager start --standalone` to start the selenium server I get the following error 

```
Error: getaddrinfo ENOTFOUND selenium-release.storage.googleapis.com selenium-release.storage.googleapis.com:443
    at errnoException (dns.js:50:10)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:92:26)
```


which I guess is similar to the one being discussed in this thread. Is there any solution to this yet? I am sorry but the cross linking of issues is a bit confusing and I have read all the linked but unable to determine if there is a way to start the webdriver manager offline or not. Don't expect a fix(if not available) but please tell me if its possible or not.

**Comment by light24bulbs** [comment]
We need a way to run the webdriver on servers that do not have internet access.  It's more common than you would think in high-security environments.  A very common use case would be to run update when building the CiCd jenkins container, than then do 
`webdriver-manager start --offline` if such a command existed.  
This feature is needed.  `start` currently violates the separation of concerns.

**Comment by light24bulbs** [comment]
Just to reiterate since this seems to be getting no attention, the following command works fine in my firewalled CI environment.
`java -Dwebdriver.chrome.driver=/usr/lib/node_modules/protractor/node_modules/webdriver-manager/selenium/chromedriver_2.39 -Dwebdriver.gecko.driver=/usr/lib/node_modules/protractor/node_modules/webdriver-manager/selenium/geckodriver-v0.20.1 -jar /usr/lib/node_modules/protractor/node_modules/webdriver-manager/selenium/selenium-server-standalone-3.12.0.jar -port 4444
`
Which is all webdriver-manager invokes, but `webdriver-manager run` hangs for about a minute and then fails in a weird way.  That's not a good behavior.  People will not be expecting that command to try to access the internet, and if it does need to, they would expect it to at least failover gracefully and run locally.

**Pull request merged**